File: frontend/lib/br-validation.ts
const digitsOnly = (value: string) => value.replace(/\D/g, "");

export function isValidEmail(value: string): boolean {
  const text = value.trim();
  if (!text || text.length > 254) return false;
  return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(text);
}

export function isValidPhoneBr(value: string): boolean {
  const d = digitsOnly(value);
  return d.length === 10 || d.length === 11;
}

function cpfChecksum(nums: number[], factor: number): number {
  const total = nums.reduce((sum, n, i) => sum + n * (factor - i), 0);
  const rest = (total * 10) % 11;
  return rest === 10 ? 0 : rest;
}

export function isValidCpf(value: string): boolean {
  const d = digitsOnly(value);
  if (d.length !== 11 || /^(\d)\1{10}$/.test(d)) return false;
  const nums = d.split("").map(Number);
  if (cpfChecksum(nums.slice(0, 9), 10) !== nums[9]) return false;
  if (cpfChecksum(nums.slice(0, 10), 11) !== nums[10]) return false;
  return true;
}

function cnpjChecksum(nums: number[], weights: number[]): number {
  const total = nums.reduce((sum, n, i) => sum + n * weights[i], 0);
  const rest = total % 11;
  return rest < 2 ? 0 : 11 - rest;
}

export function isValidCnpj(value: string): boolean {
  const d = digitsOnly(value);
  if (d.length !== 14 || /^(\d)\1{13}$/.test(d)) return false;
  const nums = d.split("").map(Number);
  const w1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
  const w2 = [6, ...w1];
  if (cnpjChecksum(nums.slice(0, 12), w1) !== nums[12]) return false;
  if (cnpjChecksum(nums.slice(0, 13), w2) !== nums[13]) return false;
  return true;
}

export function isValidCpfOrCnpj(value: string): boolean {
  const d = digitsOnly(value);
  if (d.length === 11) return isValidCpf(d);
  if (d.length === 14) return isValidCnpj(d);
  return false;
}

export function validationMessageForPerson(document: string, email: string, phone: string): string | null {
  if (!isValidCpfOrCnpj(document)) return "CPF ou CNPJ inválido.";
  if (!isValidEmail(email)) return "E-mail inválido.";
  if (!isValidPhoneBr(phone)) return "Telefone inválido (10 ou 11 dígitos).";
  return null;
}

File: backend/app/br_validation.py
"""Validação de documentos e contatos (Brasil)."""

from __future__ import annotations

import re

_DIGITS = re.compile(r"\D+")


def only_digits(value: str | None) -> str:
    return _DIGITS.sub("", value or "")


def is_valid_email(value: str | None) -> bool:
    text = (value or "").strip()
    if not text or len(text) > 254:
        return False
    return bool(re.match(r"^[^@\s]+@[^@\s]+\.[^@\s]+$", text))


def is_valid_phone_br(value: str | None) -> bool:
    digits = only_digits(value)
    return len(digits) in {10, 11}


def _cpf_checksum(digits: list[int], factor: int) -> int:
    total = sum(n * f for n, f in zip(digits, range(factor, 1, -1), strict=False))
    rest = (total * 10) % 11
    return 0 if rest == 10 else rest


def is_valid_cpf(value: str | None) -> bool:
    digits = only_digits(value)
    if len(digits) != 11 or digits == digits[0] * 11:
        return False
    nums = [int(c) for c in digits]
    if _cpf_checksum(nums[:9], 10) != nums[9]:
        return False
    if _cpf_checksum(nums[:10], 11) != nums[10]:
        return False
    return True


def _cnpj_checksum(digits: list[int], weights: list[int]) -> int:
    total = sum(n * w for n, w in zip(digits, weights, strict=False))
    rest = total % 11
    return 0 if rest < 2 else 11 - rest


def is_valid_cnpj(value: str | None) -> bool:
    digits = only_digits(value)
    if len(digits) != 14 or digits == digits[0] * 14:
        return False
    nums = [int(c) for c in digits]
    w1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]
    w2 = [6] + w1
    if _cnpj_checksum(nums[:12], w1) != nums[12]:
        return False
    if _cnpj_checksum(nums[:13], w2) != nums[13]:
        return False
    return True


def is_valid_cpf_or_cnpj(value: str | None) -> bool:
    digits = only_digits(value)
    if len(digits) == 11:
        return is_valid_cpf(digits)
    if len(digits) == 14:
        return is_valid_cnpj(digits)
    return False


def contact_validation_errors(
    *,
    document: str | None,
    email: str | None,
    phone: str | None,
    person_type: str | None = "PF",
) -> list[str]:
    errors: list[str] = []
    person = (person_type or "PF").upper()
    digits = only_digits(document)
    if person == "PJ":
        if not is_valid_cnpj(digits):
            errors.append("CNPJ inválido.")
    else:
        if not is_valid_cpf(digits):
            errors.append("CPF inválido.")
    if not is_valid_email(email):
        errors.append("E-mail inválido.")
    if not is_valid_phone_br(phone):
        errors.append("Telefone inválido (10 ou 11 dígitos).")
    return errors


def assert_valid_contact(
    *,
    document: str | None,
    email: str | None,
    phone: str | None,
    person_type: str | None = "PF",
) -> None:
    from fastapi import HTTPException

    errors = contact_validation_errors(
        document=document, email=email, phone=phone, person_type=person_type
    )
    if errors:
        raise HTTPException(status_code=422, detail=errors[0])
